File: src/services/settingsSlug.ts
/**
 * 設定ファイル名の slug 化と衝突解決 (#913)。
 *
 * 不変条件: 正規アイテムのファイル basename は「slugify の不動点」
 * (小文字 `[a-z0-9-]`・48 文字上限・Windows 予約デバイス名でない)。
 * 規約適合の判定 (isSlugConforming) と生成 (slugifyName) は
 * 同じ変換を共有し、生成物が常に適合するよう保証する。
 *
 * 占有判定の casefold・探索空間 (対応表 ∪ ディレクトリ実列挙 ∪ ID 集合)
 * は呼び出し側の責務。ここは純関数のみ。
 */

export const SLUG_MAX_LENGTH = 48

/**
 * Windows 予約デバイス名。ファイル名の stem がこれに一致すると
 * Windows で作成・削除不能になるため、slug として出力しない。
 */
const RESERVED_DEVICE_NAMES = new Set([
  'con',
  'prn',
  'aux',
  'nul',
  ...Array.from({ length: 9 }, (_, i) => `com${i + 1}`),
  ...Array.from({ length: 9 }, (_, i) => `lpt${i + 1}`),
])

/** ASCII case-insensitive 照合用の正規化。 */
export function casefold(s: string): string {
  return s.toLowerCase()
}

/** 末尾ハイフンの除去 (切り詰めで生じた分)。 */
function trimTrailingHyphens(s: string): string {
  return s.replace(/-+$/, '')
}

/** 予約名なら回避 suffix を付ける。上限内に収まるよう調整。 */
function avoidReserved(slug: string): string {
  if (!RESERVED_DEVICE_NAMES.has(slug)) return slug
  return `${slug}-x`
}

/**
 * 表示名を slug 化する。空になったら fallback (種別名) に落ちる。
 * 出力は常に isSlugConforming を満たす不動点。
 */
export function slugifyName(name: string, fallback: string): string {
  let slug = name
    .toLowerCase()
    .replace(/[^a-z0-9-]+/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-+/, '')
  slug = trimTrailingHyphens(slug)
  if (slug.length > SLUG_MAX_LENGTH) {
    slug = trimTrailingHyphens(slug.slice(0, SLUG_MAX_LENGTH))
  }
  if (!slug) slug = fallback
  return avoidReserved(slug)
}

/**
 * basename が規約適合か = slugify の不動点か。
 * 移行 (a) の正規化対象判定・新規書込の検証に使う。
 */
export function isSlugConforming(basename: string): boolean {
  if (!basename) return false
  // fallback を経由しない純変換との一致で判定する
  if (!/^[a-z0-9-]+$/.test(basename)) return false
  if (basename.length > SLUG_MAX_LENGTH) return false
  if (/--/.test(basename)) return false
  if (basename.startsWith('-') || basename.endsWith('-')) return false
  if (RESERVED_DEVICE_NAMES.has(basename)) return false
  return true
}

/**
 * base に連番 suffix を合成する。上限超過時は base 側を切り詰め、
 * 切断で生じた末尾ハイフンを除去して不動点を保つ。
 */
export function composeSuffixed(base: string, n: number): string {
  const suffix = `-${n}`
  let head = base
  if (head.length + suffix.length > SLUG_MAX_LENGTH) {
    head = trimTrailingHyphens(head.slice(0, SLUG_MAX_LENGTH - suffix.length))
  }
  return `${head}${suffix}`
}

/**
 * 空き名を探索する。base が占有されていれば `-2` から昇順連番。
 * isTaken には「対応表 ∪ ディレクトリ実列挙 ∪ (ID を決める操作では)
 * ID 集合」に対する casefold 済みの述語を渡すこと。
 */
export function resolveAvailable(
  base: string,
  isTaken: (candidate: string) => boolean,
): string {
  if (!isTaken(base)) return base
  for (let n = 2; ; n++) {
    const candidate = composeSuffixed(base, n)
    if (!isTaken(candidate)) return candidate
  }
}

File: crates/notecore/src/settings_slug.rs
//! 設定ファイル名 (slug) の規則 (#913)。TS の `src/services/settingsSlug.ts` と同じ
//! 規則で、collection (skill / テーマ / プラグイン等) のファイル名を決める。

pub const SLUG_MAX_LENGTH: usize = 48;

const RESERVED: &[&str] = &[
    "con", "prn", "aux", "nul", "com1", "com2", "com3", "com4", "com5", "com6", "com7", "com8",
    "com9", "lpt1", "lpt2", "lpt3", "lpt4", "lpt5", "lpt6", "lpt7", "lpt8", "lpt9",
];

fn is_reserved(s: &str) -> bool {
    RESERVED.contains(&s)
}

fn collapse_dashes(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut prev_dash = false;
    for c in s.chars() {
        if c == '-' {
            if !prev_dash {
                out.push('-');
            }
            prev_dash = true;
        } else {
            out.push(c);
            prev_dash = false;
        }
    }
    out
}

/// 表示名 → slug。小文字化し、`[a-z0-9-]` 以外の連続を `-` に、連続する `-` を
/// 1 つに、両端の `-` を落とし、上限で切る。空なら fallback、予約名なら `-x`。
pub fn slugify_name(name: &str, fallback: &str) -> String {
    let lower = name.to_lowercase();
    let mut replaced = String::with_capacity(lower.len());
    let mut in_run = false;
    for c in lower.chars() {
        if c.is_ascii_lowercase() || c.is_ascii_digit() || c == '-' {
            replaced.push(c);
            in_run = false;
        } else if !in_run {
            replaced.push('-');
            in_run = true;
        }
    }
    let mut slug = collapse_dashes(&replaced);
    slug = slug.trim_matches('-').to_string();
    if slug.len() > SLUG_MAX_LENGTH {
        slug.truncate(SLUG_MAX_LENGTH);
        slug = slug.trim_end_matches('-').to_string();
    }
    if slug.is_empty() {
        slug = fallback.to_string();
    }
    if is_reserved(&slug) {
        slug = format!("{slug}-x");
    }
    slug
}

/// slug 規約に沿った名前か (空でない / `[a-z0-9-]` のみ / 上限内 / `--` 無し /
/// 両端が `-` でない / 予約名でない)。
pub fn is_slug_conforming(b: &str) -> bool {
    !b.is_empty()
        && b.chars()
            .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '-')
        && b.len() <= SLUG_MAX_LENGTH
        && !b.contains("--")
        && !b.starts_with('-')
        && !b.ends_with('-')
        && !is_reserved(b)
}

/// `base-n`。上限を超えるなら base を削ってから付ける。
pub fn compose_suffixed(base: &str, n: usize) -> String {
    let suffix = format!("-{n}");
    if base.len() + suffix.len() <= SLUG_MAX_LENGTH {
        return format!("{base}{suffix}");
    }
    let keep = SLUG_MAX_LENGTH - suffix.len();
    let mut head: String = base.chars().take(keep).collect();
    head = head.trim_end_matches('-').to_string();
    format!("{head}{suffix}")
}

/// 空いていればそのまま、占有されていれば `-2`, `-3`, … で最初に空くものを返す。
pub fn resolve_available(base: &str, is_taken: impl Fn(&str) -> bool) -> String {
    if !is_taken(base) {
        return base.to_string();
    }
    let mut n = 2;
    loop {
        let candidate = compose_suffixed(base, n);
        if !is_taken(&candidate) {
            return candidate;
        }
        n += 1;
    }
}

pub fn casefold(s: &str) -> String {
    s.to_lowercase()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn slugify_matches_ts_examples() {
        assert_eq!(slugify_name("My Theme", "theme"), "my-theme");
        assert_eq!(slugify_name("Weather_2", "widget"), "weather-2");
        assert_eq!(slugify_name("日本語だけ", "skill"), "skill");
        assert_eq!(slugify_name("プロファイル 1", "profile"), "1");
        assert_eq!(slugify_name("CON", "x"), "con-x");
        assert_eq!(slugify_name("auxiliary", "x"), "auxiliary");
        let long = format!("{}-bbbb", "a".repeat(47));
        assert_eq!(slugify_name(&long, "x"), "a".repeat(47));
        for s in ["My Theme", "Weather_2", "CON", &long] {
            let once = slugify_name(s, "x");
            assert_eq!(slugify_name(&once, "x"), once, "fixed point for {s}");
        }
    }

    #[test]
    fn conforming_and_suffix_rules() {
        assert!(is_slug_conforming("my-skill"));
        assert!(!is_slug_conforming("My-skill"));
        assert!(!is_slug_conforming("a--b"));
        assert!(!is_slug_conforming("-a"));
        assert!(!is_slug_conforming("con"));
        assert!(!is_slug_conforming(""));
        let taken = ["foo", "foo-2", "foo-3"];
        assert_eq!(resolve_available("foo", |c| taken.contains(&c)), "foo-4");
        let s = compose_suffixed(&"a".repeat(48), 2);
        assert!(s.len() <= 48 && s.ends_with("-2"));
    }
}
